core/kernel/ldelf_syscalls.c

```diff
@@ -38,26 +96,25 @@ void ta_bin_close(void *ptr)
 	free(binh);
 }
 
-TEE_Result ldelf_open_ta_binary(struct system_ctx *ctx, uint32_t param_types,
-				TEE_Param params[TEE_NUM_PARAMS])
+TEE_Result ldelf_sc_open_bin(const TEE_UUID *uuid, size_t uuid_size,
+			     uint32_t *handle)
 {
 	TEE_Result res = TEE_SUCCESS;
+	struct ts_session *sess = ts_get_current_session();
+	struct system_ctx **sys_ctx = (struct system_ctx **)&sess->user_ctx;
+	struct system_ctx *ctx = NULL;
 	struct bin_handle *binh = NULL;
-	int h = 0;
-	TEE_UUID *uuid = NULL;
 	uint8_t tag[FILE_TAG_SIZE] = { 0 };
 	unsigned int tag_len = sizeof(tag);
-	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
-					  TEE_PARAM_TYPE_VALUE_OUTPUT,
-					  TEE_PARAM_TYPE_NONE,
-					  TEE_PARAM_TYPE_NONE);
+	int h = 0;
 
-	if (exp_pt != param_types)
-		return TEE_ERROR_BAD_PARAMETERS;
-	if (params[0].memref.size != sizeof(*uuid))
+	if (uuid_size != sizeof(*uuid))
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	uuid = params[0].memref.buffer;
+	ctx = calloc(1, sizeof(*ctx));
+	if (!ctx)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	*sys_ctx = ctx;
```

Comment:
Ack for first 2 points.

> Will this be freed if ldelf crashes or otherwise the session is closed before this has been freed?

I think not, AFAICS this was already a problem before the PR. If there is an error in ldelf, we return with `_utee_panic()` or through the `err()` macro which invokes `_utee_return()`. But neither of these seem to close the system PTA session, which would free the `sys_ctx`. Or am I missing something? Do you have a suggestion on this?

Reply:
> Ack for first 2 points.
> 
> > Will this be freed if ldelf crashes or otherwise the session is closed before this has been freed?
> 
> I think not, AFAICS this was already a problem before the PR. If there is an error in ldelf, we return with `_utee_panic()` or through the `err()` macro which invokes `_utee_return()`. But neither of these seems to close the system PTA session, which would free the `sys_ctx`. Or am I missing something? Do you have a suggestion on this?

Previously it was cleaned up as part of closing the session. In this case we could take care of that in `free_utc()` for user TAs and the counterpart for SPs. Except that we don't have access to the session pointer there. Perhaps `user_ta_enter_close_session()` instead?

Reply:
I see, `user_ta_enter_close_session()` is a good idea.
I'm also looking at functions in `ldelf_loader.c` as a possible candidate instead. What do you think about doing the cleanup there, right after returning from `thread_enter_user_mode()`, if ldelf panicked or returned with an error?

Reply:
@jenswi-linaro: Could you share your opinion on this proposed solution?

Reply:
> I'm also looking at functions in `ldelf_loader.c` as a possible candidate instead. What do you think about doing the cleanup there, right after returning from `thread_enter_user_mode()`, if ldelf panicked or returned with an error?

Sure, that should work also. We should perhaps always cleanup there since ldelf is done for the moment and shouldn't keep any handles open.

Reply:
Thanks, done.